File: 124-binary-tree-maximum-path-sum/binary-tree-maximum-path-sum.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function maxPathSum(root: TreeNode | null): number {
    let maxSum: number = -Infinity

    function dfs(node: TreeNode | null): number {
        if (node === null) return 0
        
        const leftMax: number = Math.max(dfs(node.left), 0)
        const rightMax: number = Math.max(dfs(node.right), 0)

        const currentSum: number = node.val + leftMax + rightMax

        maxSum = Math.max(currentSum, maxSum)

        return node.val + Math.max(leftMax, rightMax)
    }

    dfs(root)

    return maxSum
};

File: 124-binary-tree-maximum-path-sum/binary-tree-maximum-path-sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        max_path_sum = float('-inf')
        def dfs(node):
            if not node:
                return 0
            nonlocal max_path_sum

            left_max = max(dfs(node.left), 0)
            right_max = max(dfs(node.right), 0)

            max_path_sum = max(max_path_sum, left_max + right_max + node.val)
            
            return max(left_max, right_max) + node.val

        dfs(root)
        return max_path_sum